File: sdks/android/src/main/kotlin/app/tracki/sdk/Assist.kt
package app.tracki.sdk

internal data class AssistContentWire(
    val title: String,
    val body: String,
    val cta: CtaContent?,
) {
    companion object {
        @Suppress("UNCHECKED_CAST")
        fun from(j: Map<String, Any?>?): AssistContentWire {
            val m = j ?: emptyMap()
            val c = m["cta"] as? Map<String, Any?>
            val cta = c?.let {
                CtaContent(
                    label = it["label"]?.toString() ?: "",
                    kind = it["kind"] as? String,
                    url = it["url"] as? String,
                    faq = it["faq"] as? Boolean,
                )
            }
            return AssistContentWire(
                title = m["title"]?.toString() ?: "",
                body = m["body"]?.toString() ?: "",
                cta = cta,
            )
        }
    }
}

internal data class AssistPayload(
    val actionId: String,
    val mode: String,
    val arContent: AssistContentWire,
    val enContent: AssistContentWire,
    val arArticle: Pair<String, String>?,
    val enArticle: Pair<String, String>?,
    val articleId: String?,
) {
    companion object {
        @Suppress("UNCHECKED_CAST")
        fun from(raw: Any?): AssistPayload? {
            val m = raw as? Map<String, Any?> ?: return null
            val actionId = m["actionId"] as? String ?: return null
            val content = m["content"] as? Map<String, Any?> ?: return null

            fun article(a: Any?): Pair<String, String>? {
                val d = a as? Map<String, Any?> ?: return null
                return (d["title"]?.toString() ?: "") to (d["body"]?.toString() ?: "")
            }
            val art = m["article"] as? Map<String, Any?>

            return AssistPayload(
                actionId = actionId,
                mode = m["mode"] as? String ?: "fallback",
                arContent = AssistContentWire.from(content["ar"] as? Map<String, Any?>),
                enContent = AssistContentWire.from(content["en"] as? Map<String, Any?>),
                arArticle = article(art?.get("ar")),
                enArticle = article(art?.get("en")),
                articleId = m["articleId"] as? String,
            )
        }
    }
}

/**
 * Server-driven Live Assist, mobile edition (slice 5 mechanism unchanged):
 * the worker matched a struggle to an action + FAQ and stashed the payload;
 * it arrives on an event-flush response and renders as a contextual drawer.
 * Once per session per action; fail-silent.
 */
class AssistHandler(
    private val locale: () -> String,
    private val currentPath: () -> String,
    private val queue: EventQueue,
    private val renderer: Renderer?,
    private val cta: CtaRouter,
    private val now: Clock,
) {
    private val shown = HashSet<String>()

    private fun emit(type: String, p: AssistPayload) {
        val props = linkedMapOf<String, Any?>("action_id" to p.actionId, "mode" to p.mode)
        if (p.articleId != null) props["article_id"] = p.articleId
        queue.enqueue(EventInput(type = type, ts = now(), path = currentPath(), props = props))
    }

    internal fun show(payload: AssistPayload) {
        val r = renderer ?: return
        if (shown.contains(payload.actionId)) return
        val L = locale()
        val authored = if (L == "en") payload.enContent else payload.arContent
        var title = authored.title
        var body = authored.body
        if (payload.mode == "answer" && (payload.arArticle != null || payload.enArticle != null)) {
            val candidate = if (L == "en") payload.enArticle else payload.arArticle
            val a = if (candidate != null && candidate.first.isNotEmpty()) candidate
            else (payload.arArticle ?: payload.enArticle)
            if (a != null) {
                title = a.first
                body = a.second
            }
        }
        if (title.isEmpty() && body.isEmpty()) return
        shown.add(payload.actionId)

        val ctaContent = authored.cta
        val intent = AssistIntent(
            actionId = payload.actionId,
            mode = payload.mode,
            title = title,
            body = body,
            ctaLabel = ctaContent?.label,
            helpful = { emit("assist_helpful", payload) },
            unhelpful = {
                emit("assist_unhelpful", payload)
                // Parity with the snippet: an unhelpful answer offers the Agent.
                cta.openChat()
            },
            escalate = {
                emit("assist_escalate", payload)
                if (ctaContent != null) {
                    val kind = ctaKind(ctaContent)
                    when {
                        kind == CtaKind.URL -> cta.activate(ctaContent)
                        kind == CtaKind.FAQ && ctaContent.kind == "faq" -> cta.openFaq()
                        kind == CtaKind.WHATSAPP -> cta.openWhatsApp()
                        else -> // chat — and the legacy slice-5 shape (faq:true,
                            // no kind), which keeps its audited semantics of
                            // escalating to the Agent.
                            cta.openChat()
                    }
                }
            },
        )
        try {
            r.show(intent)
            emit("assist_shown", payload)
        } catch (e: Throwable) {
            // fail-silent
        }
    }

    /** Wire as the queue's response handler. */
    @Suppress("UNCHECKED_CAST")
    fun onResponse(data: Any?) {
        val raw = (data as? Map<String, Any?>)?.get("assist") ?: return
        AssistPayload.from(raw)?.let { show(it) }
    }
}

File: sdks/flutter/lib/src/assist.dart
import 'cta.dart';
import 'queue.dart';
import 'types.dart';

class _AssistContentWire {
  final String title;
  final String body;
  final CtaContent? cta;
  const _AssistContentWire({required this.title, required this.body, this.cta});

  static _AssistContentWire fromJson(Map j) {
    CtaContent? cta;
    final c = j['cta'];
    if (c is Map) {
      cta = CtaContent(
        label: '${c['label'] ?? ''}',
        kind: c['kind'] as String?,
        url: c['url'] as String?,
        faq: c['faq'] is bool ? c['faq'] as bool : null,
      );
    }
    return _AssistContentWire(
      title: '${j['title'] ?? ''}',
      body: '${j['body'] ?? ''}',
      cta: cta,
    );
  }
}

class _AssistPayload {
  final String actionId;
  final String mode; // "answer" | "fallback"
  final _AssistContentWire arContent;
  final _AssistContentWire enContent;
  final Map<String, String>? arArticle;
  final Map<String, String>? enArticle;
  final String? articleId;

  const _AssistPayload({
    required this.actionId,
    required this.mode,
    required this.arContent,
    required this.enContent,
    this.arArticle,
    this.enArticle,
    this.articleId,
  });

  static _AssistPayload? fromJson(Object? raw) {
    if (raw is! Map) return null;
    final actionId = raw['actionId'];
    final content = raw['content'];
    if (actionId is! String || content is! Map) return null;
    Map<String, String>? parseArticle(Object? a) {
      if (a is! Map) return null;
      return {'title': '${a['title'] ?? ''}', 'body': '${a['body'] ?? ''}'};
    }

    final article = raw['article'];
    return _AssistPayload(
      actionId: actionId,
      mode: '${raw['mode'] ?? 'fallback'}',
      arContent: _AssistContentWire.fromJson((content['ar'] as Map?) ?? const {}),
      enContent: _AssistContentWire.fromJson((content['en'] as Map?) ?? const {}),
      arArticle: (article is Map) ? parseArticle(article['ar']) : null,
      enArticle: (article is Map) ? parseArticle(article['en']) : null,
      articleId: raw['articleId'] as String?,
    );
  }
}

/// Server-driven Live Assist, mobile edition (slice 5 mechanism unchanged):
/// the worker matched a struggle to an action + FAQ and stashed the payload;
/// it arrives on an event-flush response and renders as a contextual drawer.
/// Once per session per action; fail-silent.
class AssistHandler {
  final String Function() locale;
  final String Function() currentPath;
  final EventQueue queue;
  final Renderer? renderer;
  final CtaRouter cta;
  final Clock now;

  final Set<String> _shown = {};

  AssistHandler({
    required this.locale,
    required this.currentPath,
    required this.queue,
    required this.renderer,
    required this.cta,
    required this.now,
  });

  void _emit(String type, _AssistPayload p) {
    final props = <String, Object?>{'action_id': p.actionId, 'mode': p.mode};
    if (p.articleId != null) props['article_id'] = p.articleId;
    queue.enqueue(EventInput(
      type: type,
      ts: now(),
      path: currentPath(),
      props: props,
    ));
  }

  void show(_AssistPayload payload) {
    final r = renderer;
    if (r == null || _shown.contains(payload.actionId)) return;
    final L = locale();
    final authored = L == 'en' ? payload.enContent : payload.arContent;
    var title = authored.title;
    var body = authored.body;
    if (payload.mode == 'answer' && (payload.arArticle != null || payload.enArticle != null)) {
      final candidate = L == 'en' ? payload.enArticle : payload.arArticle;
      final a = (candidate != null && (candidate['title'] ?? '').isNotEmpty)
          ? candidate
          : (payload.arArticle ?? payload.enArticle);
      if (a != null) {
        title = a['title'] ?? '';
        body = a['body'] ?? '';
      }
    }
    if (title.isEmpty && body.isEmpty) return;
    _shown.add(payload.actionId);

    final ctaContent = authored.cta;
    final intent = AssistIntent(
      actionId: payload.actionId,
      mode: payload.mode,
      title: title,
      body: body,
      ctaLabel: ctaContent?.label,
      helpful: () => _emit('assist_helpful', payload),
      unhelpful: () {
        _emit('assist_unhelpful', payload);
        // Parity with the snippet: an unhelpful answer offers the Agent.
        cta.openChat();
      },
      escalate: () {
        _emit('assist_escalate', payload);
        if (ctaContent != null) {
          final kind = ctaKind(ctaContent);
          if (kind == CtaKind.url) {
            cta.activate(ctaContent);
          } else if (kind == CtaKind.faq && ctaContent.kind == 'faq') {
            cta.openFaq();
          } else if (kind == CtaKind.whatsapp) {
            cta.openWhatsApp();
          } else {
            // chat — and the legacy slice-5 shape (faq:true, no kind), which
            // keeps its audited semantics of escalating to the Agent.
            cta.openChat();
          }
        }
      },
    );
    try {
      r.show(intent);
      _emit('assist_shown', payload);
    } catch (_) {
      // fail-silent
    }
  }

  /// Wire as the queue's response handler.
  void onResponse(Object? data) {
    final raw = (data is Map) ? data['assist'] : null;
    if (raw == null) return;
    final assist = _AssistPayload.fromJson(raw);
    if (assist != null) show(assist);
  }
}
